benchmarks/benchmark-merkle-trees.ts

```diff
@@ -0,0 +1,340 @@
+    b.suite(
+        name,
+        b.add(`IMT - Add ${numberOfLeaves} leaves`, () => {
+            try {
+                if (leafIMT < numberOfLeaves) {
+                    incrementalMerkleTree2.insert(leafIMT)
+                    leafIMT += 1
+                }
+            } catch (error) {
+                logger.error(name)
+                logger.error(error)
+            }
+        }),
+        b.add(`LeanIMT - Add ${numberOfLeaves} leaves`, () => {
+            try {
+                leanIncrementalMerkleTree2.insert(BigInt(leafLeanIMT))
+            } catch (error) {
+                logger.error(name)
+                logger.error(error)
+            }
+            leafLeanIMT += 1
+        }),
+        b.add(`SparseMT - Add ${numberOfLeaves} leaves`, () => {
+            try {
+                sparseMerkleTree2.add(leafSMT.toString(16), Math.floor(Math.random() * 10).toString())
+            } catch (error) {
+                logger.error(name)
+                logger.error(error)
+            }
+            leafSMT += 1
+        }),
+        b.cycle(),
+        b.complete(),
+        b.save({ folder: "benchmarks/results", file: name, version: "1.0.0", details: true }),
+        b.save({ folder: "benchmarks/results", file: name, format: "chart.html", details: true }),
+        b.save({ folder: "benchmarks/results", file: name, format: "table.html", details: true })
+    ).catch((error) => {
+        logger.error(name)
+        logger.error(error)
+    })
+
+    /*  Benchmarking Proof Generation:
+            This section benchmarks proof generation for each tree type. */
+    leafIMT = 0
+    leafLeanIMT = 0
+    leafSMT = 0
+    name = `proof-generation-merkle-trees-${numberOfLeaves}`
+    b.suite(
+        name,
+        b.add(`IMT - Generated ${samples} proofs`, () => {
+            try {
+                if (leafIMT < samples) {
+                    incrementalMerkleTree.createProof(sample[leafIMT])
+                } else {
+                    leafIMT = 0
+                }
+                leafIMT += 1
+            } catch (error) {
+                logger.error(name)
+                logger.error(error)
+            }
+        }),
+        b.add(`LeanIMT - Generated ${samples} proofs`, () => {
+            try {
+                if (leafLeanIMT < samples) {
+                    leanIncrementalMerkleTree.generateProof(sample[leafLeanIMT])
+                } else {
+                    leafLeanIMT = 0
+                }
+                leafLeanIMT += 1
+            } catch (error) {
+                logger.error(name)
+                logger.error(error)
+            }
+        }),
+        b.add(`SparseMT - Generated ${samples} proofs`, () => {
+            try {
+                if (leafSMT < samples) {
+                    sparseMerkleTree.createProof(sample[leafSMT].toString(16))
+                } else {
+                    leafSMT = 0
+                }
+                leafSMT += 1
```

Comment:
I am not sure I am understand how the leaf index/count is supposed behave.
If leaf count === samples, leaf count will be set to 1 (first 0 then increased by 1). Is it what we want?

Reply:
**Benny does not have a setting to limit the number of cycles** will be performed then I put this code to restart doing the operations from the first leaf, giving the user the possibility to use the same benchmark to test trees of any sizes without unnecessary errors (index overflow) during the benchmark execution.

Reply:
You right, I have changed it in the last commit